### ISSUE TITLE
Consider adding headings to allowed content in <label> element

### ISSUE THREAD

**Issue by 36degrees** [w3c/html#1270]
In #724 the spec was changed to allow headings as permitted children of legends, but this same change was not also extended to labels.

This means that [including a heading as part of a legend](https://codepen.io/36degrees/pen/ZrwRQM) is valid, but [including a heading as part of a label](https://codepen.io/36degrees/pen/rJPKeQ) is not.

Is there any reason why the label should not behave the same as the legend in this regard?

I am happy to put together a more detailed case for this change, and test behaviour in browsers and assistive technologies, but wanted to open an issue early to make sure there wasn't something obvious I'd missed that would prevent this from even being an option.

Thanks.

**Comment by chaals** [comment]
I think explaining the use case (i.e. concrete examples where you would use this) and testing that it doesn't break stuff are the first steps. If you are happy to do that, I don't know of any reason to hold back.

(Something to test is what happens, e.g. in a screen reader, when you have a heading in a label, and how it interacts  with ARIA headings...)

**Comment by scottaohara** [comment]
Agreed with @chaals.  I'd be interested to understand more about why this would be a necessary use case and why, if a `label` did in fact need to double as a heading, the following valid markup pattern wouldn't suffice:

```
<h3>
  <label for="input">
    test
  </label>
</h3>
<input type="text" id="input">
```

**Comment by 36degrees** [comment]
Thanks, both!

The specific use case we're trying to cater for here is where the user is presented with a page with a single question on it, where the question is both the heading for the page and also the label/legend for the field/fieldset.

For example, in this example, the question 'What is your date of birth?' is marked up as a legend with a heading inside it. This was [a change we made in June 2017 after a lot of research by a colleague](https://github.com/alphagov/govuk_elements/pull/507).

![govuk-design-system-production cloudapps digital_patterns_question-pages_default_index html](https://user-images.githubusercontent.com/121939/36851054-3aac8eaa-1d60-11e8-9210-d6993cd5331e.png)

This works great, but only because of the change made in #724 which permits headings to exist within legends.

Of course, if the question does not require a fieldset then instead we need to make the heading a label:

![govuk-design-system-production cloudapps digital_patterns_question-pages_postcode_index html](https://user-images.githubusercontent.com/121939/36851127-6aa18a52-1d60-11e8-87f4-442e65f830a0.png)

In this case, our _current approach_ is exactly what @scottaohara suggested – rather than putting the heading inside the label, we invert it and put the label inside the heading.

However, we've identified two issues with this:

Firstly, we are currently re-implementing this as part of a design system where everything is styled using explicit classes and we're finding that the label class is overriding the heading styling, and so the resulting element looks like a label:
```
<h1 class="govuk-heading-xl"><label class="govuk-c-label">test</label></h1>
```

I recognise that this is not insurmountable – one possible solution we've come up with is to provide a modifier for the label that makes it look like a heading:

```
<h1><label class="govuk-c-label govuk-c-label--heading-xl">test</label></h1>
```

However, this introduces extra complexity for developers where I think it _could_ be avoided. Allowing headings inside labels would allow us to use the same approach in both cases, rather than developers having to remember that the heading should go inside a legend with the normal heading class, but outside a label _without a class_ and _with a modifier_ on the label. I strongly suspect that what will happen is that a lot of developers will end up putting the headings inside the `<label>` anyway, because it works and that's what they.

Secondly, where the field has a 'hint' or an error message, we associate them with the field by including them inside the label. This means that the hint and error message also become part of the H1 in this case. This isn't an issue with the fieldset, where the hint and error message are inside the label but outside of the H1. We are currently [exploring whether we can move these out of the label and associate them with `aria-describedby` instead](https://github.com/alphagov/govuk_elements/issues/574), which may make this a non-issue, but we need to do further testing to make sure this isn't going to case any issues with assistive technologies.

**Comment by scottaohara** [comment]
@36degrees, if I'm understanding you correctly, this seems to be more of a specificity issue with your design system than a markup issue, or problem with the spec?

I know I don't have full context here, but I would not expect a label to also be a page heading.  

I also wouldn't expect a label to have a label class on it, if it were meant to look like a heading it was contained within.  If everything is styled via classes, then having the heading with its appropriate class, and a label without a class applied to it, should then fix the styling specificity issue you're encountering?

Here is a [codepen in how I understand your issue](https://codepen.io/scottohara/pen/aqMoeO) and outlining how I'd expect to mitigate this.

**Comment by 36degrees** [comment]
>  if I'm understanding you correctly, this seems to be more of a specificity issue with your design system than a markup issue, or problem with the spec?

Yep, that's a fair enough comment. But to me it feels like it's a _workaround_ and that in this case the spec might not be right.

In #724, the history of the legend element was reported as:

> This history FWIW is that HTML4 restricted legend to inline/phrasing and we just propagated that because as far as I remember at least nobody til now ever proposed loosening it.

I would like to know if the same holds true for labels – is it just that this is a hangover from HTML4 and no-one has ever proposed loosening it? If so, I think there is merit in making labels and legends behave in a similar way.

Conversely, if this is by design – if there is a good reason why why legends can contain headings but labels cannot, then it would be really useful to understand what that reason is.

> I know I don't have full context here, but I would not expect a label to also be a page heading.

I think the second screenshot above, which demonstrates our 'one thing per page' format is a good example of a scenario where you would want to do exactly that. It was previously implemented as a visually hidden H1 with a visible label, but after a lot of testing this was changed as a result of feedback from our accessibility community – as described in some detail [here](https://github.com/alphagov/govuk_elements/pull/507).

**Comment by scottaohara** [comment]
Appreciate the additional context @36degrees.

Definitely something to discuss further.  Thanks again for raising the issue.

**Comment by 36degrees** [comment]
Hi,

I've built a couple of basic test cases for the ['current implementation' (label inside heading)](https://36degrees.github.io/headings-inside-labels/inverse) and the [proposal (heading inside label)](https://36degrees.github.io/headings-inside-labels/) and used these to capture screenshots and test with assistive technologies.

I haven't found to cause concern – it seems to behave as expected with everything I tested.

## Browser Screenshots

<details>
<summary>Android Google Nexus 9</summary>

## Heading inside label (proposed behaviour)
![android_google-nexus-9_5 0_portrait_heading_inside_label](https://user-images.githubusercontent.com/121939/37915874-66ddc672-3112-11e8-9d83-c9af6c49d7a9.png)

## Label inside heading (current behaviour)
![android_google-nexus-9_5 0_portrait_label_inside_heading](https://user-images.githubusercontent.com/121939/37915875-66f6e1b6-3112-11e8-9399-61e743f3ca5b.png)
</details>

<details>
<summary>Android Samsung Galaxy S5 </summary>

## Heading inside label (proposed behaviour)
![android_samsung-galaxy-s5_4 4_portrait_heading_inside_label](https://user-images.githubusercontent.com/121939/37915876-670f90da-3112-11e8-9e47-83d72cf054be.png)

## Label inside heading (current behaviour)
![android_samsung-galaxy-s5_4 4_portrait_label_inside_heading](https://user-images.githubusercontent.com/121939/37915877-672c0936-3112-11e8-880b-c1c782824b7e.png)
</details>

<details>
<summary>iPad 3rd Generation (iOS 6.0)</summary>

## Heading inside label (proposed behaviour)
![ios_ipad-3rd- 6 0 _6 0_portrait_heading_inside_label](https://user-images.githubusercontent.com/121939/37915879-6742e00c-3112-11e8-9995-4e01079764aa.png)

## Label inside heading (current behaviour)
![ios_ipad-3rd- 6 0 _6 0_portrait_label_inside_heading](https://user-images.githubusercontent.com/121939/37915881-67bd0c88-3112-11e8-863d-b08459cd2c67.png)
</details>

<details>
<summary>iPad Mini (iOS 7.0)</summary>

## Heading inside label (proposed behaviour)
![ios_ipad-mini_7 0_portrait_heading_inside_label](https://user-images.githubusercontent.com/121939/37915882-67d5246c-3112-11e8-987e-443836555384.png)

## Label inside heading (current behaviour)
![ios_ipad-mini_7 0_portrait_label_inside_heading](https://user-images.githubusercontent.com/121939/37915884-67ee6b02-3112-11e8-9586-262fe87a9d0f.png)
</details>

<details>
<summary>iPad Mini 2 (iOS 8.3)</summary>

## Heading inside label (proposed behaviour)
![ios_ipad-mini-2_8 3_portrait_heading_inside_label](https://user-images.githubusercontent.com/121939/37915885-680a1104-3112-11e8-8ce3-2e23870d237b.png)

## Label inside heading (current behaviour)
![ios_ipad-mini-2_8 3_portrait_label_inside_heading](https://user-images.githubusercontent.com/121939/37915886-685921c2-3112-11e8-981f-2fcc8f4263d4.png)
</details>

<details>
<summary>Safari 9.1 (Mac OS X El Capitan)</summary>

## Heading inside label (proposed behaviour)
![macelc_safari_9 1_heading_inside_label](https://user-images.githubusercontent.com/121939/37915887-6870f4dc-3112-11e8-9fcf-7bba7a73e116.png)

## Label inside heading (current behaviour)
![macelc_safari_9 1_label_inside_heading](https://user-images.githubusercontent.com/121939/37915888-68865e3a-3112-11e8-8972-0483a2cb290d.png)
</details>

<details>
<summary>Safari 8.0 (Mac OS X Yosemite)</summary>

## Heading inside label (proposed behaviour)
![macyos_safari_8 0_heading_inside_label](https://user-images.githubusercontent.com/121939/37915889-689e2c86-3112-11e8-806f-b6d01f58246c.png)

## Label inside heading (current behaviour)
![macyos_safari_8 0_label_inside_heading](https://user-images.githubusercontent.com/121939/37915890-68b5eb5a-3112-11e8-9b49-00889b34aeb8.png)
</details>

<details>
<summary>IE8 (Windows 7)</summary>

## Heading inside label (proposed behaviour)
![win7_ie_8 0_heading_inside_label](https://user-images.githubusercontent.com/121939/37915891-68cebdb0-3112-11e8-896a-ffd47b0b023e.png)

## Label inside heading (current behaviour)
![win7_ie_8 0_label_inside_heading](https://user-images.githubusercontent.com/121939/37915892-68f0147e-3112-11e8-824b-031b20e5337f.png)
</details>

<details>
<summary>IE9 (Windows 7)</summary>

## Heading inside label (proposed behaviour)
![win7_ie_9 0_heading_inside_label](https://user-images.githubusercontent.com/121939/37915893-694cd0ce-3112-11e8-9660-6b821e544d7a.png)

## Label inside heading (current behaviour)
![win7_ie_9 0_label_inside_heading](https://user-images.githubusercontent.com/121939/37915894-69843334-3112-11e8-8a01-965be8ae4d89.png)
</details>

<details>
<summary>IE10 (Windows 8)</summary>

## Heading inside label (proposed behaviour)
![win8_ie_10 0_heading_inside_label](https://user-images.githubusercontent.com/121939/37915895-69a03df4-3112-11e8-81d7-1d605273db3a.png)

## Label inside heading (current behaviour)
![win8_ie_10 0_label_inside_heading](https://user-images.githubusercontent.com/121939/37915896-69b8f4ca-3112-11e8-8a57-d57aa1b46eaf.png)
</details>

<details>
<summary>IE11 (Windows 8)</summary>

## Heading inside label (proposed behaviour)
![win8 1_ie_11 0_heading_inside_label](https://user-images.githubusercontent.com/121939/37915897-69d18f1c-3112-11e8-8bda-73396b0c27f6.png)

## Label inside heading (current behaviour)
![win8 1_ie_11 0_label_inside_heading](https://user-images.githubusercontent.com/121939/37915899-69ec7084-3112-11e8-9e2b-2c35ec2bad9b.png)
</details>

<details>
<summary>Chrome 48 (Windows 10)</summary>

## Heading inside label (proposed behaviour)
![win10_chrome_48 0_heading_inside_label](https://user-images.githubusercontent.com/121939/37915900-6a07de32-3112-11e8-828e-db95fa5b1604.png)

## Label inside heading (current behaviour)
![win10_chrome_48 0_label_inside_heading](https://user-images.githubusercontent.com/121939/37915902-6a2533ba-3112-11e8-9a1c-24a35fa3cdd7.png)
</details>

<details>
<summary>Chrome 49 (Windows 10)</summary>

## Heading inside label (proposed behaviour)
![win10_chrome_49 0_heading_inside_label](https://user-images.githubusercontent.com/121939/37915903-6a48bf10-3112-11e8-9092-837914a77503.png)

## Label inside heading (current behaviour)
![win10_chrome_49 0_label_inside_heading](https://user-images.githubusercontent.com/121939/37915904-6a61c19a-3112-11e8-940a-73aa83d47879.png)
</details>

<details>
<summary>Chrome 50 (Windows 10)</summary>

## Heading inside label (proposed behaviour)
![win10_chrome_50 0_heading_inside_label](https://user-images.githubusercontent.com/121939/37915905-6a7e3ca8-3112-11e8-87a2-c581b4856ee4.png)

## Label inside heading (current behaviour)
![win10_chrome_50 0_label_inside_heading](https://user-images.githubusercontent.com/121939/37915907-6abdb6bc-3112-11e8-94cf-50b32aee6829.png)
</details>

<details>
<summary>Edge 15 (Windows 10)</summary>

## Heading inside label (proposed behaviour)
![win10_edge_15 0_heading_inside_label](https://user-images.githubusercontent.com/121939/37915908-6ad910d8-3112-11e8-90b4-2b5431a816d1.png)

## Label inside heading (current behaviour)
![win10_edge_15 0_label_inside_heading](https://user-images.githubusercontent.com/121939/37915909-6b052416-3112-11e8-986c-efebe4fb6753.png)
</details>

<details>
<summary>Firefox 43 (Windows 10)</summary>

## Heading inside label (proposed behaviour)
![win10_firefox_43 0_heading_inside_label](https://user-images.githubusercontent.com/121939/37915910-6b23a97c-3112-11e8-8934-5e7c41ebd9bc.png)

## Label inside heading (current behaviour)
![win10_firefox_43 0_label_inside_heading](https://user-images.githubusercontent.com/121939/37915912-6b555c4c-3112-11e8-8b43-596deda3e7a8.png)
</details>

<details>
<summary>Firefox 44 (Windows 10)</summary>

## Heading inside label (proposed behaviour)
![win10_firefox_44 0_heading_inside_label](https://user-images.githubusercontent.com/121939/37915913-6b6feed6-3112-11e8-9e8e-9b98f54e82c1.png)

## Label inside heading (current behaviour)
![win10_firefox_44 0_label_inside_heading](https://user-images.githubusercontent.com/121939/37915914-6b8cc31c-3112-11e8-835a-9ea5eb371dd9.png)
</details>

<details>
<summary>Firefox 45 (Windows 10)</summary>

## Heading inside label (proposed behaviour)
![win10_firefox_45 0_heading_inside_label](https://user-images.githubusercontent.com/121939/37915915-6ba5588c-3112-11e8-8ef1-ade6bcd4c7ca.png)

## Label inside heading (current behaviour)
![win10_firefox_45 0_label_inside_heading](https://user-images.githubusercontent.com/121939/37915916-6bbd1a1c-3112-11e8-95d7-861fb03f77de.png)
</details>

## Assistive technologies

I've tested with the [combinations of assistive technologies and browsers as recommended by the GOV.UK Service Manual](https://www.gov.uk/service-manual/technology/testing-with-assistive-technologies#what-to-test) – I'm happy to test further combinations if it helps to build confidence in this change.

<details>
  <summary>JAWS 17, IE11, Windows 10</h2></summary>

---

<strong>On page load</strong>

Headings inside labels test case, Headings inside labels test case, Headings inside labels test case

Page has one heading and no links. Headings inside labels test case, Heading level 1: What is your home postcode? Edit, continue button, continue"

(This behaviour is the same with the current implementation ('inverse' test case), including the duplication of the page title etc.

<strong>Using list of all headings (JAWS + F6)</strong>

JAWS lists one heading, What is your home postcode? : 1, 1 of 1.

[Selecting the heading, JAWS announces:]

"Heading level 1, what is your home postcode? what is your home postcode? Heading level 1"

(This duplication of the heading happens in other places – with the existing implementation and e.g. the H1 on the GOV.UK homepage)

<strong>Moving through headings using 'h' shortcut</strong>

JAWS focusses the heading and announces "What is your home postcode, heading level 1"

[Tab to field from there:]

What is your home postcode? edit, type of text.
</details>

<details>
  <summary>ZoomText 10.1, IE11, Windows 10</summary>
  
---

  <strong>Using Reader speech mode</strong>

Clicking in the text field, ZoomText announces "What is your home postcode?"

<strong>Using AppRdr</strong>

Clicking on the start of the heading, ZoomText announces "What is your home postcode? Continue, end of document".

This is the same behaviour as with the inverse test case.

</details>

<details>
  <summary>Dragon NaturallySpeaking 13, IE11, Windows 10</summary>

---

<strong>Able to focus and fill in form field using voice prompts:</strong>

- "Click what is your home postcode?"
- "Click postcode"
</details>

<details>
  <summary>NVDA 2017.4, Mozilla Firefox 52.7.2, Windows 10</summary>
  
---

Clickable heading level 1, what is your home postcode

(The current implementation reads "Heading level 1 clickable, what is your home postcode")

Edit has autocomplete, button continue

Navigating backwards from the input using Shift-H:

"Clickable what is your home postcode, heading level 1"

(The current implementation reads "Clickable what is your home postcode, heading level 1")

Navigating through form fields using F:

"What is your home postcode, edit has autocomplete"
</details>

<details>
  <summary>VoiceOver with Safari 11, OS El Capitan</summary>

---

<strong>Navigating through the page from the top</strong>

Voiceover announces "Start interacting with Headings inside labels test case, HTML content, What is your home postcode? group"

[Navigating to the next element using VO + Right arrow]

Voiceover announces "What is your home postcode? edit text"

[Navigating back to the previous heading using VO + Cmd + Shift + H]

Voiceover announces "Heading level 1, What is your home postcode?"

<strong>Using the rotor</strong>

Heading "What is your home post code?" appears in headings list in rotor

Navigating to heading using the rotor focusses the cursor on the heading

Form control "What is your home post code? edit text" appears in form control list in rotor.

Navigating to form control from the rotor focusses the form field and VoiceOver announces "What is your home postcode? edit text"
</details>


<details>
  <summary>VoiceOver, Safari with iOS 11.2.6 (iPhone X)</summary>

---

[Swiping right from top of page until the form field is focussed]

Voiceover announces "What is your home postcode? Text field. Double tap to edit."

[Headings selected in rotor menu.]

Voiceover announces "1 heading".

From top of document, "What is your home postcode? Heading level 1"
</details>

**Comment by 36degrees** [comment]
What are the next steps that need to be done to make progress on this?

**Comment by chaals** [comment]
@36degrees 

> What are the next steps that need to be done to make progress on this?

At a minimum, the ping you just made. More effective would be writing a Pull Request that made the change to the spec.

The logic and the use case seem reasonable, and it is hard to imagine what this would actually break... 

Ping @sideshowbarker

**Comment by 36degrees** [comment]
Closed based on discussion in #1304:

> [Due to the issues I've outlined I do not think this change should be accepted without much more consideration and tests that provide evidence this is a needed change beyond this specific use case.](https://github.com/w3c/html/pull/1304#pullrequestreview-108490156)
